### PR TITLE
Buffered serialization

### PR DESCRIPTION
**Pull request by benjaminwinger**
Splits the SerDeser class into a Serializer and Deserializer, and a BufferedFileWriter (for serializing to disk), and BufferedSerializer (for serializing to an in-memory buffer).

### CODE REVIEW

File: src/include/common/serializer/writer.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+#include <cstdint>
+
+namespace kuzu {
+namespace common {
+
+class Writer {
+public:
+    virtual void write(const uint8_t* data, uint64_t size) = 0;
+    virtual ~Writer(){};
```
Comment (acquamarin) [inline src/include/common/serializer/writer.h:11]: I think our convension is to use default for trivial deconstructors.
virtual ~Writer() = default;

**Review by acquamarin — APPROVED** [review]

File: src/common/serializer/deserializer.cpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#include "common/serializer/deserializer.h"
+
+namespace kuzu {
+namespace common {
+
+template<>
+void Deserializer::deserializeValue<std::string>(std::string& value) {
```
Comment (acquamarin) [inline src/common/serializer/deserializer.cpp:7]: I think you don't need to write the specialization explicitly in the implementation, this should work:
`void Deserializer::deserializeValue(std::string& value)`

**Review by acquamarin — COMMENTED** [review]
BTW: We should also refactor parquetWrtier to use the file_writer.

**Comment by benjaminwinger** [comment]
> BTW: We should also refactor parquetWrtier to use the file_writer.

Can that be left to a later PR, or do you want it included in this one?

**Pull request merged**